### PR TITLE
Validation loop

### PR DESCRIPTION
**Pull request by FilyaGeikyan**
You can choose a validation dataset.
The workflow logs some metrics like this:
[rank0]:2024-09-05 16:03:33,336 - root - INFO - context: val  step: 1  loss: 11.2918  memory:  1.59GiB(16.28%)  wps: 106,703  mfu: 1.92%
More metrics are shown in aim, just like training.

### CODE REVIEW

File: torchtitan/config_manager.py
Comment (philippguevorguian) [inline torchtitan/config_manager.py:1]: Can we run a simple check in the validate_config function that ensures the following when validation is enabled:
1. the validation dataset path has been provided **and **exists.
2. eval frequency has been provided.

File: torchtitan/config_manager.py [hunk 1]
```diff
@@ -369,6 +370,27 @@ def __init__(self):
             help="Python garbage control scheduling interval, in steps",
         )
 
+        # validation configs
+        self.parser.add_argument(
+            "--validation.batch_size", type=int, default=0
```
Comment (philippguevorguian) [inline torchtitan/config_manager.py:375]: The default for validation batch size should be the training batch size. Here let's have the default as None, and add a help discription that explains it defaults to the training batch size when not passed.

File: train.py [hunk 2]
```diff
@@ -99,9 +100,12 @@ def main(job_config: JobConfig):
         job_config.training.batch_size,
         job_config.training.seq_len,
         dp_degree,
-        dp_rank,
+        dp_rank
     )
 
+    # validation batch size
+    val_bs = job_config.validation.batch_size if job_config.validation.batch_size != 0 else job_config.metrics.batch_size
```
Comment (philippguevorguian) [inline train.py:107]: See above comment, if the default is None this line becomes more readable
`val_bs = job_config.validation.batch_size if job_config.validation.batch_size else job_config.metrics.batch_size`

File: train_configs/debug_model.toml [hunk 3]
```diff
@@ -38,13 +38,19 @@ gradient_accumulation_steps = 1
...
 tensor_parallel_degree = 1
 compile = false
 dataset = "chemlactica_train_mini"  # supported datasets: c4_test (2K), c4 (177M), chemlactica_train_mini (4K)
 data_process_style="chemlactica_style"
 
+[validation]
+batch_size = 1
+eval_freq = 3
+enable_val = true
+dataset = "c4_test"  # supported datasets: c4_test (2K), c4 (177M)
```
Comment (philippguevorguian) [inline train_configs/debug_model.toml:52]: looks like dataset works? If we can use both datasets and dataset paths, let's make sure only one of the two is passed in validate_config.

File: torchtitan/val.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
+import time
+from torch.nn.functional import cross_entropy
+
+def loss_fn(pred, labels):
```
Comment (philippguevorguian) [inline torchtitan/val.py:4]: This function should be the same as what is used in train, separately specifying it here can lead to issues if we make changes. Let's pull the train loss_fn and this one into a separate file and import it in both files.

File: train.py [hunk 5]
```diff
@@ -364,6 +371,47 @@ def loss_fn(pred, labels):
...
+                        job_config.training.seq_len,
+                    )
+                    validate(
+                        model,
+                        val_data_loader,
+                        eval_state,
+                        logger,
+                        metric_logger,
+                        parallel_dims, 
+                        gpu_memory_monitor,
```
Comment (philippguevorguian) [inline train.py:404]: This is a lot of arguments, but most are needed. It would be nice to wrap this in a dataclass.

File: train.py [hunk 6]
```diff
@@ -364,6 +371,47 @@ def loss_fn(pred, labels):
...
+                    num_flop_per_token_val = utils.get_num_flop_per_token_forward(
+                        utils.get_num_params(model, exclude_embedding=True),
+                        model_config,
+                        job_config.training.seq_len,
+                    )
+                    validate(
+                        model,
+                        val_data_loader,
+                        eval_state,
+                        logger,
```
Comment (philippguevorguian) [inline train.py:401]: Passing the logger isn't necessary, simply initialize the logger with the init_logger function in val.py

**Review by philippguevorguian — CHANGES_REQUESTED** [review]
Couple of points mainly related to file organization and configuration checks.

File: train_configs/debug_model.toml (hunk 3 above)
Comment (FilyaGeikyan) [reply train_configs/debug_model.toml:52]: > looks like dataset works? If we can use both datasets and dataset paths, let's make sure only one of the two is passed in validate_config.

the dataset_path is found in the build_hf_... so we need to pass both and it automatically finds the path, if You were talking about that

Comment (philippguevorguian) [reply train_configs/debug_model.toml:52]: That's right, I didn't realize this was the original intended functionality. Let's keep it the way it is.

**Review by philippguevorguian — DISMISSED** [review]
LGTM

**Comment by tigranfah** [comment]
What was the reason that it did not work back then?

**Comment by tigranfah** [comment]
overall the implementation makes sense, but it is not cleaned up, the merging with the main branch might not be correct.

**Review by tigranfah — DISMISSED** [review]

**Review by philippguevorguian — CHANGES_REQUESTED** [review]
Changes:
- move config validation
- correct clipping location
- verify hash change

File: train.py [hunk 7]
```diff
@@ -187,9 +199,6 @@ def loss_fn(pred, labels):
 
     train_state = TrainState()
 
-    metric_logger = build_metric_logger(job_config, parallel_dims)
-    metric_logger.log_hparams(job_config.args_dict)
```
[the remark this answers is not in the document]
Comment (philippguevorguian) [reply train.py:191]: This was moved down, it's still there

File: train.py [hunk 8]
```diff
@@ -27,31 +25,18 @@
...
-            yield
-
-    return context
+from torchtitan.val import validate
+from typing import Any
 
 
 # Enable debug tracing on failure: https://pytorch.org/docs/stable/elastic/errors.html
 @record
-def main(job_config: JobConfig):
```
[the remark this answers is not in the document]
Comment (philippguevorguian) [reply train.py:54]: This eventually needs to be changed thoroughly via more robust typing, but currently LSP complains about all attribute accesses into job_config object, because the attributes are set dynamically in config_manager. I think it's better to keep it as any, since this interface is really simple and the warnings from LSP are misleading.

File: torchtitan/config_manager.py [hunk 9]
```diff
@@ -569,13 +589,13 @@ def __init__(self):
...
             type=int,
-            default=300,
+            default=30000,
             help="Timeout for communication operations, during initialization and first train step.",
         )
         self.parser.add_argument(
             "--comm.train_timeout_seconds",
             type=int,
-            default=100,
+            default=10000,
```
[the remark this answers is not in the document]
Comment (philippguevorguian) [reply torchtitan/config_manager.py:598]: Agree we should change this back

File: torchtitan/val.py [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+        del loss
+
+    if os.path.exists(end_of_validation_path) and dp_rank == 0:
+        os.remove(end_of_validation_path)
+        logger.info("removed the store file")
+    else:
+        logger.info("no store file exists")
+    del val_data_iterator, data_loader, end_of_validation_store
+
+    gc.collect()
```
[the remark this answers is not in the document]
Comment (philippguevorguian) [reply torchtitan/val.py:160]: gc_handler will not run a collection unless it reaches a certain number of iterations. I think it's better to ensure we collect at the end of evaluation

File: torchtitan/val.py [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+        enable_compiled_autograd,
+    )
+
+    process_ids = [str(rank) for rank in range(world_size)]
+
+    model.eval()
+
+    eval_state.step = 0
+    val_data_iterator = iter(data_loader)
+    while True:
```
[the remark this answers is not in the document]
Comment (philippguevorguian) [reply torchtitan/val.py:79]: We can do this if it doesn't slow down performance.

File: train.py [hunk 12]
```diff
@@ -255,43 +270,44 @@ def loss_fn(pred, labels):
...
                     # need to free to before bwd to avoid peaking memory
                     del pred
                     loss.backward()
-                
-                for m in model_parts:
-                    torch.nn.utils.clip_grad_norm_(
-                        m.parameters(), job_config.training.max_norm, foreach=True
-                    )
-                    
+            logger.debug(f"after backward {train_state.step}")
```
[the remark this answers is not in the document]
Comment (philippguevorguian) [reply train.py:304]: These aren't visible unless we set log level to debug, but I am not opinionated on this. @FilyaGeikyan let's remove the new debug statements.

File: train.py [hunk 13]
```diff
@@ -392,6 +413,42 @@ def loss_fn(pred, labels):
                 time_last_log = time.perf_counter()
                 gpu_memory_monitor.reset_peak_stats()
 
+            # log val metrics
+            if job_config.validation.enable_val and (
+                train_state.step == 0
+                or train_state.step % eval_freq == 0
+            ):
+                val_data_loader = build_hf_data_loader(
```
[the remark this answers is not in the document]
Comment (philippguevorguian) [reply train.py:421]: Simplifies the interface at neglible performance cost, since we do all data processing on-the-fly. Otherwise, we will have to add a method to reset the dataloader state. Feel free to add this if you're willing to.

File: train.py [hunk 14]
```diff
@@ -198,7 +207,6 @@ def loss_fn(pred, labels):
         lr_schedulers=lr_schedulers.schedulers,
         states={"train_state": train_state},
         job_config=job_config,
-        experiment_hash=metric_logger.experiment_hash
```
[the remark this answers is not in the document]
Comment (philippguevorguian) [reply train.py:201]: I agree, @FilyaGeikyan  let's bring this back.

File: torchtitan/val.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (tigranfah) [reply torchtitan/val.py:79]: my point is that we have a unified interface for gc, why not use that in all places consistently

File: train.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (tigranfah) [reply train.py:191]: 1. it should be before CheckpointManager to pass experiment hash to aim
2. the other implementation is wrong

I believe this happened because of incorrect merge with current main branch

File: test/assets/chemlactica_valid_mini/file0.jsonl
Comment (philippguevorguian) [inline test/assets/chemlactica_valid_mini/file0.jsonl:1]: Did we test this with multiple files in validation?

File: torchtitan/val.py [hunk 15]
```diff
@@ -150,12 +161,12 @@ def validate(
     if loss:
         del loss
 
-    if os.path.exists(end_of_validation_path) and dp_rank == 0:
-        os.remove(end_of_validation_path)
-        logger.info("removed the store file")
-    else:
-        logger.info("no store file exists")
-    del val_data_iterator, data_loader, end_of_validation_store
+    # if os.path.exists(end_of_validation_path) and dp_rank == 0:
```
Comment (philippguevorguian) [inline torchtitan/val.py:164]: If we aren't using the store for signalling end of validation, we can remove these comments

File: torchtitan/val.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (philippguevorguian) [reply torchtitan/val.py:81]: This is implemented now.

File: train.py [hunk 16]
```diff
@@ -92,23 +92,33 @@ def main(job_config: Any):
...
+        dp_rank,
+        False,
+        pin_memory=job_config.dataloader.pin_memory,
+        num_workers=job_config.dataloader.num_workers,
+        special_mode=job_config.dataloader.special_mode,
+    )
+
     # check the validation parameters
-    if not job_config.validation.dataset_path and not job_config.validation.dataset:
+    if not job_config.validation.dataset:
```
Comment (philippguevorguian) [inline train.py:116]: let's move this to the config_manager and only raise the error if validation is enabled.

**Review by philippguevorguian — COMMENTED** [review]
Looks great, concern about testing with multiple validation files and not needing to raise errors from validation dataset configuration if validation is not enabled.

File: test/assets/chemlactica_valid_mini/file0.jsonl
Comment (tigranfah) [reply test/assets/chemlactica_valid_mini/file0.jsonl:1]: we have single validation file

**Comment by tigranfah** [comment]
as a final test, i have the chemlactica-125m reproduction run with validation turned on, and it looks good, I think it is safe to merge. Reference run (green), run with validation on (blue).
Screenshot 2024-09-28 at 11 36 25 AM

**Comment by tigranfah** [comment]
@philippguevorguian @FilyaGeikyan

**Review by philippguevorguian — APPROVED** [review]
Code LGTM, thanks for providing the loss curves as well
